File: scripts/fix-stuck-sprint.py
#!/usr/bin/env python3
"""
fix-stuck-sprint.py  —  hard reset of the active sprint tables.

Clears EVERYTHING from `sprints` and `sprint_stories`.
Sprint history (sprint_history table) is NOT touched.
User stories are NOT deleted — just set back to 'completed'.

Usage:
    python scripts/fix-stuck-sprint.py <project-path>

Example:
    python scripts/fix-stuck-sprint.py C:\\Users\\jjdub\\code\\baanbaan\\Merchant
"""

import sys, os, sqlite3
from datetime import datetime, timezone

if len(sys.argv) < 2:
    print("Usage: python fix-stuck-sprint.py <project-path>")
    sys.exit(1)

project_path = sys.argv[1]
puffin_dir   = os.path.join(project_path, ".puffin")
db_path      = os.path.join(puffin_dir, "puffin.db")
json_path    = os.path.join(puffin_dir, "active-sprint.json")

if not os.path.exists(db_path):
    print(f"Database not found: {db_path}")
    sys.exit(1)

print(f"Opening: {db_path}\n")
con = sqlite3.connect(db_path)
con.row_factory = sqlite3.Row

# --- Show current state ---
sprints = con.execute("SELECT id, title, status, closed_at FROM sprints ORDER BY created_at").fetchall()
print(f"sprints table ({len(sprints)} rows):")
for s in sprints:
    print(f"  status={s['status']:<14} closed_at={repr(s['closed_at']):<32} \"{s['title']}\"")

links = con.execute("SELECT COUNT(*) FROM sprint_stories").fetchone()[0]
print(f"sprint_stories table: {links} rows\n")

if not sprints:
    print("sprints table is already empty.")
else:
    print("This will DELETE all rows from `sprints` and `sprint_stories`.")
    print("Sprint history (sprint_history) and user stories are NOT affected.\n")
    answer = input("Proceed? (yes/no): ").strip().lower()
    if answer != "yes":
        print("Aborted.")
        con.close()
        sys.exit(0)

    now = datetime.now(timezone.utc).isoformat()
    con.execute("PRAGMA foreign_keys = OFF")

    try:
        con.execute("BEGIN IMMEDIATE")
        con.execute("DELETE FROM sprint_stories")
        con.execute("DELETE FROM sprints")
        # Return any in-progress stories to pending so they show up in the backlog
        result = con.execute(
            "UPDATE user_stories SET status='pending', updated_at=? WHERE status='in-progress'",
            (now,)
        )
        con.execute("COMMIT")
        print(f"Cleared sprints table.")
        print(f"Cleared sprint_stories table.")
        if result.rowcount:
            print(f"Reset {result.rowcount} in-progress stories → pending.")
    except Exception as e:
        con.execute("ROLLBACK")
        print(f"Error (rolled back): {e}")
        con.close()
        sys.exit(1)

    con.execute("PRAGMA foreign_keys = ON")

# --- Delete JSON backup so migrator can't re-populate on restart ---
if os.path.exists(json_path):
    os.remove(json_path)
    print(f"Deleted backup: {json_path}")

# --- Final check ---
remaining = con.execute("SELECT COUNT(*) FROM sprints").fetchone()[0]
print(f"\nRows remaining in sprints: {remaining}")
history = con.execute("SELECT COUNT(*) FROM sprint_history").fetchone()[0]
print(f"Sprint history entries preserved: {history}")
print("\nDone. Restart Puffin — sprint history will still be there for Rerun.")
con.close()

File: scripts/fix-stuck-sprint.js
#!/usr/bin/env node
/**
 * fix-stuck-sprint.js
 *
 * Removes the stuck active sprint from a Puffin project database and
 * returns its user stories to 'completed' status.
 *
 * Usage:
 *   node fix-stuck-sprint.js <project-path>
 *
 * Example:
 *   node fix-stuck-sprint.js C:\Users\jjdub\code\baanbaan\Merchant
 */

const path = require('path')
const fs = require('fs')

// --- Resolve project path ---
const projectPath = process.argv[2]
if (!projectPath) {
  console.error('Usage: node fix-stuck-sprint.js <project-path>')
  process.exit(1)
}

const dbPath = path.join(projectPath, '.puffin', 'puffin.db')
if (!fs.existsSync(dbPath)) {
  console.error(`Database not found at: ${dbPath}`)
  process.exit(1)
}

// --- Load better-sqlite3 from Puffin's own node_modules ---
const scriptDir = path.dirname(require.resolve('./fix-stuck-sprint.js') || __filename)
const puffinRoot = path.resolve(scriptDir, '..')
const sqlitePath = path.join(puffinRoot, 'node_modules', 'better-sqlite3')
let Database
try {
  Database = require(sqlitePath)
} catch (e) {
  console.error(`Could not load better-sqlite3 from ${sqlitePath}`)
  console.error('Run this script from the puffin project root, or install better-sqlite3 globally.')
  process.exit(1)
}

const db = new Database(dbPath)
db.pragma('journal_mode = WAL')
db.pragma('foreign_keys = ON')

// --- Find the stuck active sprint ---
const activeSprint = db.prepare(`
  SELECT * FROM sprints WHERE closed_at IS NULL ORDER BY created_at DESC LIMIT 1
`).get()

if (!activeSprint) {
  console.log('No active sprint found — database is already clean.')
  db.close()
  process.exit(0)
}

console.log(`Found active sprint:`)
console.log(`  ID:     ${activeSprint.id}`)
console.log(`  Title:  ${activeSprint.title}`)
console.log(`  Status: ${activeSprint.status}`)
console.log(`  Created: ${activeSprint.created_at}`)

// --- Get linked story IDs ---
const storyRows = db.prepare(`
  SELECT story_id FROM sprint_stories WHERE sprint_id = ?
`).all(activeSprint.id)

const storyIds = storyRows.map(r => r.story_id)
console.log(`\nLinked stories: ${storyIds.length}`)

if (storyIds.length > 0) {
  // Check current story statuses
  const stories = db.prepare(`
    SELECT id, title, status FROM user_stories WHERE id IN (${storyIds.map(() => '?').join(',')})
  `).all(...storyIds)

  stories.forEach(s => console.log(`  [${s.status.padEnd(12)}] ${s.title}`))
}

// --- Confirm before proceeding ---
const readline = require('readline')
const rl = readline.createInterface({ input: process.stdin, output: process.stdout })

rl.question(`\nThis will:\n  1. Delete sprint "${activeSprint.title}"\n  2. Set its ${storyIds.length} stories to 'completed'\n  3. Remove sprint_stories links\n\nProceed? (yes/no): `, (answer) => {
  rl.close()

  if (answer.toLowerCase() !== 'yes') {
    console.log('Aborted.')
    db.close()
    process.exit(0)
  }

  // --- Execute in a transaction ---
  const fix = db.transaction(() => {
    // 1. Return stories to 'completed'
    if (storyIds.length > 0) {
      const now = new Date().toISOString()
      const updateStmt = db.prepare(`
        UPDATE user_stories SET status = 'completed', updated_at = ? WHERE id = ?
      `)
      for (const id of storyIds) {
        updateStmt.run(now, id)
      }
      console.log(`\nSet ${storyIds.length} stories → completed`)
    }

    // 2. Remove sprint_stories links
    const deletedLinks = db.prepare(`
      DELETE FROM sprint_stories WHERE sprint_id = ?
    `).run(activeSprint.id)
    console.log(`Removed ${deletedLinks.changes} sprint_stories links`)

    // 3. Delete the sprint
    db.prepare(`DELETE FROM sprints WHERE id = ?`).run(activeSprint.id)
    console.log(`Deleted sprint: ${activeSprint.title}`)
  })

  fix()

  // --- Verify ---
  const remaining = db.prepare(`SELECT COUNT(*) as n FROM sprints WHERE closed_at IS NULL`).get()
  console.log(`\nActive sprints remaining: ${remaining.n}`)
  console.log('Done. Restart Puffin to pick up the changes.')

  db.close()
})
